java/source/flight.rst

```diff
@@ -0,0 +1,1050 @@
+            @Override
+            public Runnable acceptPut(CallContext context, FlightStream flightStream, StreamListener<PutResult> ackStream) {
+                List<ArrowRecordBatch> listArrowRecordBatch = new ArrayList<>();
+                return () -> {
+                    long rows = 0;
+                    while (flightStream.next()) {
+                        VectorUnloader unloader = new VectorUnloader(flightStream.getRoot());
+                        try (final ArrowRecordBatch arb = unloader.getRecordBatch()) {
+                            listArrowRecordBatch.add(arb);
+                            rows += flightStream.getRoot().getRowCount();
+                        }
+                    }
+                    DataInMemory pojoFlightDataInMemory = new DataInMemory(listArrowRecordBatch, flightStream.getSchema(), rows);
+                    dataInMemory.put(flightStream.getDescriptor(), pojoFlightDataInMemory);
+                    ackStream.onCompleted();
+                };
+            }
+
+            @Override
+            public void getStream(CallContext context, Ticket ticket, ServerStreamListener listener) {
+                FlightDescriptor flightDescriptor = FlightDescriptor.path(new String(ticket.getBytes(), StandardCharsets.UTF_8)); // Recover data for key configured
+                if(dataInMemory.containsKey(flightDescriptor)){
+                    VectorSchemaRoot vectorSchemaRoot = VectorSchemaRoot.create(dataInMemory.get(flightDescriptor).getSchema(), allocator);
+                    listener.start(vectorSchemaRoot);
+                    for(ArrowRecordBatch arrowRecordBatch : dataInMemory.get(flightDescriptor).getListArrowRecordBatch()){
+                        VectorLoader loader = new VectorLoader(vectorSchemaRoot);
+                        loader.load(arrowRecordBatch.cloneWithTransfer(allocator));
+                        listener.putNext();
+                    }
+                    listener.completed();
+                }
+            }
+
+            @Override
+            public void doAction(CallContext context, Action action, StreamListener<Result> listener) {
+                FlightDescriptor flightDescriptor = FlightDescriptor.path(new String(action.getBody(), StandardCharsets.UTF_8)); // For recover data for key configured
+                if(dataInMemory.containsKey(flightDescriptor)) {
+                    switch (action.getType()) {
+                        case "DELETE":
+                            dataInMemory.remove(flightDescriptor);
+                            Result result = new Result("Delete completed".getBytes(StandardCharsets.UTF_8));
+                            listener.onNext(result);
+                    }
+                    listener.onCompleted();
+                }
+            }
+
+            @Override
+            public FlightInfo getFlightInfo(CallContext context, FlightDescriptor descriptor) {
+                if(!dataInMemory.containsKey(descriptor)){
```

Comment:
Similarly we should get the value and handle the error if it doesn't exist.

Reply:
ConcurrentMap does nothing if the key is not in the map

Reply:
What do you mean?

We cannot use containsKey, it isn't so useful with ConcurrentMap. If another thread adds or removes the key in between the check, then we will get unexpected errors.

Reply:
Ok, updated